modules/assim.sequential/R/SDA_NA_downscale.R

```diff
@@ -0,0 +1,459 @@
+#' @return character: path to the exported GeoTIFF file.
+#' 
+#' @export
+#' @author Dongchen Zhang
+Average.ERA5.2.GeoTIFF <- function (start.date, end.date, in.path, outdir) {
+  # open ERA5 nc file as geotiff format for referencing crs and ext.
+  ERA5.tiff <- terra::rast(file.path(in.path, paste0("ERA5_", lubridate::year(start.date), ".nc")))
+  dates <- seq(start.date, end.date, "1 year")
+  # initialize final outcomes.
+  temp.all <- precip.all <- srd.all <- dewpoint.all <- c()
+  for (i in seq_along(dates)) {
+    # initialize start and end dates for the current period
+    if (i == 1) {
+      start <- start.date
+    } else {
+      start <- as.Date(paste0(lubridate::year(dates[i]), "-01-01"))
+    }
+    if (i == length(dates)) {
+      end <- end.date
+    } else {
+      end <- as.Date(paste0(lubridate::year(dates[i]), "-12-31"))
+    }
+    # loop over years.
+    for (j in seq_along(dates)) {
+      # open ERA5 nc file.
+      met.nc <- ncdf4::nc_open(file.path(in.path, paste0("ERA5_", lubridate::year(dates[i]), ".nc")))
+      # find index for the date.
+      times <- as.POSIXct(met.nc$dim$time$vals*3600, origin="1900-01-01 00:00:00", tz = "UTC")
+      time.inds <- which(lubridate::date(times) >= start & lubridate::date(times) <= end)
+      # extract temperature.
+      PEcAn.logger::logger.info("entering temperature.")
+      temp.all <- abind::abind(temp.all, apply(ncdf4::ncvar_get(met.nc, "t2m")[,,,time.inds], c(1,2,4), mean), along = 3)
+      # extract precipitation.
+      PEcAn.logger::logger.info("entering precipitation.")
+      precip.all <- abind::abind(precip.all, apply(ncdf4::ncvar_get(met.nc, "tp")[,,,time.inds], c(1,2,4), mean), along = 3)
+      # extract shortwave solar radiation.
+      PEcAn.logger::logger.info("entering solar radiation.")
+      srd.all <- abind::abind(srd.all, apply(ncdf4::ncvar_get(met.nc, "ssrd")[,,,time.inds], c(1,2,4), mean), along = 3)
+      # extract dewpoint.
+      PEcAn.logger::logger.info("entering dewpoint.")
+      dewpoint.all <- abind::abind(dewpoint.all, apply(ncdf4::ncvar_get(met.nc, "d2m")[,,,time.inds], c(1,2,4), mean), along = 3)
+      # close the NC connection.
+      ncdf4::nc_close(met.nc)
+    }
+  }
+  # aggregate across time.
+  # temperature.
+  temp <- apply(temp.all, c(1, 2), mean)
+  temp <- PEcAn.utils::ud_convert(temp, "K", "degC")
+  # precipitation.
+  precip <- apply(precip.all, c(1, 2), mean)
+  # solar radiation.
+  srd <- apply(srd.all, c(1, 2), mean)
+  # dewpoint.
+  dewpoint <- apply(dewpoint.all, c(1, 2), mean)
+  dewpoint <- PEcAn.utils::ud_convert(dewpoint, "K", "degC")
+  # convert dew point to relative humidity.
+  beta <- (112 - (0.1 * temp) + dewpoint) / (112 + (0.9 * temp))
+  relative.humidity <- beta ^ 8
+  VPD <- PEcAn.data.atmosphere::get.vpd(100*relative.humidity, temp)
+  # combine together.
+  PEcAn.logger::logger.info("Aggregate maps.")
+  met.rast <- c(terra::rast(matrix(temp, nrow = dim(temp)[2], ncol = dim(temp)[1], byrow = T)),
```

Comment:
If I understand correctly, the goal seems to be to generate a single map of mean climate over a specific time period. If you wanted the downscaling to use year-specific meteorology, would you call this function in a loop, iteratively calculating from July 15 of one year to July 14 of the next year, or would it make more sense to have this function generate multiple met layers at different resolutions? Similarly, if/when the SDA drops to a monthly or seasonal timestep, could this function handle calculating monthly mean weather?

Reply:
Fixed. The temporal resolution depends on the start and end dates.